[CI] CartesianShapeCentroidAggregatorTests testSingleValuedField failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/thf2rzzcleqim/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests/testSingleValuedField


**Reproduction line:**
```
./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests.testSingleValuedField" -Dtests.seed=7E2A772926C8017A -Dtests.locale=ru-RU -Dtests.timezone=America/Blanc-Sablon -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests#testSingleValuedField`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSingleValuedField'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.AssertionError: Centroid over 175 had incorrect x-value expected:<3.675715712024515E29> but was:<3.6823576417199625E29>

  at __randomizedtesting.SeedInfo.seed([7E2A772926C8017A:8A79FE88C5607F60]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:553)
  at org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests.assertCentroid(CartesianShapeCentroidAggregatorTests.java:203)
  at org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests.assertCentroid(CartesianShapeCentroidAggregatorTests.java:196)
  at org.elasticsearch.xpack.spatial.search.aggregations.metrics.CartesianShapeCentroidAggregatorTests.testSingleValuedField(CartesianShapeCentroidAggregatorTests.java:176)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
@@ -162,8 +162,9 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
                 w.addDocument(document);
                 if (targetShapeType.compareTo(calculator.getDimensionalShapeType()) == 0) {
                     double weight = calculator.sumWeight();
-                    compensatedSumLat.add(weight * calculator.getY());
-                    compensatedSumLon.add(weight * calculator.getX());
+                    // compute the centroid of centroids in float space
+                    compensatedSumLat.add(weight * (float) calculator.getY());
+                    compensatedSumLon.add(weight * (float) calculator.getX());
                     compensatedSumWeight.add(weight);
                 }
             }

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
@@ -39,7 +39,6 @@ import java.util.Collections;
 import java.util.List;
 import java.util.function.Function;
 
-import static java.lang.Math.signum;
 import static org.hamcrest.Matchers.equalTo;
 
 public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
@@ -155,8 +154,6 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
             CompensatedSum compensatedSumLon = new CompensatedSum(0, 0);
             CompensatedSum compensatedSumLat = new CompensatedSum(0, 0);
             CompensatedSum compensatedSumWeight = new CompensatedSum(0, 0);
-            TestExpected xExpected = new TestExpected();
-            TestExpected yExpected = new TestExpected();
             for (Geometry geometry : geometries) {
                 Document document = new Document();
                 CentroidCalculator calculator = new CentroidCalculator();
@@ -168,17 +165,19 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
                     compensatedSumLat.add(weight * calculator.getY());
                     compensatedSumLon.add(weight * calculator.getX());
                     compensatedSumWeight.add(weight);
-                    xExpected.setValue(compensatedSumLon.value() / compensatedSumWeight.value());
-                    yExpected.setValue(compensatedSumLat.value() / compensatedSumWeight.value());
                 }
             }
             // force using a single aggregator to compute the centroid
             w.forceMerge(1);
-            assertCentroid(w, xExpected, yExpected);
+            CartesianPoint expectedCentroid = new CartesianPoint(
+                compensatedSumLon.value() / compensatedSumWeight.value(),
+                compensatedSumLat.value() / compensatedSumWeight.value()
+            );
+            assertCentroid(w, expectedCentroid);
         }
     }
 
-    private void assertCentroid(RandomIndexWriter w, TestExpected xExpected, TestExpected yExpected) throws IOException {
+    private void assertCentroid(RandomIndexWriter w, CartesianPoint expectedCentroid) throws IOException {
         MappedFieldType fieldType = new ShapeFieldMapper.ShapeFieldType(
             "field",
             true,
@@ -194,33 +193,20 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
             assertEquals("my_agg", result.getName());
             SpatialPoint centroid = result.centroid();
             assertNotNull(centroid);
-            assertCentroid("x-value", centroid.getX(), xExpected);
-            assertCentroid("y-value", centroid.getY(), yExpected);
+            assertCentroid("x-value", result.count(), centroid.getX(), expectedCentroid.getX());
+            assertCentroid("y-value", result.count(), centroid.getY(), expectedCentroid.getY());
             assertTrue(AggregationInspectionHelper.hasValue(result));
         }
     }
 
-    private void assertCentroid(String name, double value, TestExpected expected) {
-        assertEquals("Centroid over " + expected.count + " had incorrect " + name, expected.value, value, expected.tolerance());
+    private void assertCentroid(String name, long count, double value, double expected) {
+        assertEquals("Centroid over " + count + " had incorrect " + name, expected, value, tolerance(expected, count));
     }
 
-    private static class TestExpected {
-        private double value;
-        private double previous;
-        private int count;
-
-        private void setValue(double value) {
-            this.previous = this.value;
-            this.value = value;
-            this.count++;
-        }
-
-        private double tolerance() {
-            // When last numbers change sign, they usually end up as much smaller numbers (but retain high errors)
-            double tolerance = signum(value) != signum(previous) ? Math.abs(value / 1e3) : Math.abs(value / 1e5);
-            // Very large numbers have more floating point error, also increasing with count
-            return tolerance > 1e25 ? tolerance * count : tolerance;
-        }
+    private double tolerance(double expected, long count) {
+        double tolerance = Math.abs(expected / 1e5);
+        // Very large numbers have more floating point error, also increasing with count
+        return tolerance > 1e25 ? tolerance * count : tolerance;
     }
 
     @Override

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeCentroidAggregatorTests.java
@@ -39,6 +39,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.function.Function;
 
+import static java.lang.Math.signum;
 import static org.hamcrest.Matchers.equalTo;
 
 public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
@@ -154,6 +155,8 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
             CompensatedSum compensatedSumLon = new CompensatedSum(0, 0);
             CompensatedSum compensatedSumLat = new CompensatedSum(0, 0);
             CompensatedSum compensatedSumWeight = new CompensatedSum(0, 0);
+            TestExpected xExpected = new TestExpected();
+            TestExpected yExpected = new TestExpected();
             for (Geometry geometry : geometries) {
                 Document document = new Document();
                 CentroidCalculator calculator = new CentroidCalculator();
@@ -165,19 +168,17 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
                     compensatedSumLat.add(weight * calculator.getY());
                     compensatedSumLon.add(weight * calculator.getX());
                     compensatedSumWeight.add(weight);
+                    xExpected.setValue(compensatedSumLon.value() / compensatedSumWeight.value());
+                    yExpected.setValue(compensatedSumLat.value() / compensatedSumWeight.value());
                 }
             }
             // force using a single aggregator to compute the centroid
             w.forceMerge(1);
-            CartesianPoint expectedCentroid = new CartesianPoint(
-                compensatedSumLon.value() / compensatedSumWeight.value(),
-                compensatedSumLat.value() / compensatedSumWeight.value()
-            );
-            assertCentroid(w, expectedCentroid);
+            assertCentroid(w, xExpected, yExpected);
         }
     }
 
-    private void assertCentroid(RandomIndexWriter w, CartesianPoint expectedCentroid) throws IOException {
+    private void assertCentroid(RandomIndexWriter w, TestExpected xExpected, TestExpected yExpected) throws IOException {
         MappedFieldType fieldType = new ShapeFieldMapper.ShapeFieldType(
             "field",
             true,
@@ -193,20 +194,33 @@ public class CartesianShapeCentroidAggregatorTests extends AggregatorTestCase {
             assertEquals("my_agg", result.getName());
             SpatialPoint centroid = result.centroid();
             assertNotNull(centroid);
-            assertCentroid("x-value", result.count(), centroid.getX(), expectedCentroid.getX());
-            assertCentroid("y-value", result.count(), centroid.getY(), expectedCentroid.getY());
+            assertCentroid("x-value", centroid.getX(), xExpected);
+            assertCentroid("y-value", centroid.getY(), yExpected);
             assertTrue(AggregationInspectionHelper.hasValue(result));
         }
     }
 
-    private void assertCentroid(String name, long count, double value, double expected) {
-        assertEquals("Centroid over " + count + " had incorrect " + name, expected, value, tolerance(expected, count));
+    private void assertCentroid(String name, double value, TestExpected expected) {
+        assertEquals("Centroid over " + expected.count + " had incorrect " + name, expected.value, value, expected.tolerance());
     }
 
-    private double tolerance(double expected, long count) {
-        double tolerance = Math.abs(expected / 1e5);
-        // Very large numbers have more floating point error, also increasing with count
-        return tolerance > 1e25 ? tolerance * count : tolerance;
+    private static class TestExpected {
+        private double value;
+        private double previous;
+        private int count;
+
+        private void setValue(double value) {
+            this.previous = this.value;
+            this.value = value;
+            this.count++;
+        }
+
+        private double tolerance() {
+            // When last numbers change sign, they usually end up as much smaller numbers (but retain high errors)
+            double tolerance = signum(value) != signum(previous) ? Math.abs(value / 1e3) : Math.abs(value / 1e5);
+            // Very large numbers have more floating point error, also increasing with count
+            return tolerance > 1e25 ? tolerance * count : tolerance;
+        }
     }
 
     @Override

